### PR TITLE
Issue #1191: Forbid to use unreadable characters

### LINKED ISSUE #1191: Forbid to use `lI` names
# Rule request

## Thesis

Currently one can name your classes like so: `ControlIn`
And it is unreadable! Inside it is just `L` and `i`, but we cannot tell it from the word.

We need to forbid lowercase `l` followed by uppercase `I`

## Reasoning

Readability matters!

### PR DESCRIPTION
**Pull request by Kvm99**
# I have made things!

## Checklist

- [x] I have double checked that there are no unrelated changes in this pull request (old patches, accidental config files, etc)
- [x] I have created at least one test case for the changes I have made
- [ ] I have updated the documentation for the changes I have made
- [x] I have added my changes to the `CHANGELOG.md`

## Related issues
- Closes #1191

### CODE REVIEW

File: tests/fixtures/noqa/noqa.py [hunk 1]
```diff
@@ -703,3 +703,8 @@ def consecutive_yields():
         # https://github.com/wemake-services/wemake-python-styleguide/issues/1082
         break
     my_print(4)
+
+
+class Memo0Output(object):  # noqa: WPS124
+    def __init__(self):
```
Comment (sobolevn) [inline tests/fixtures/noqa/noqa.py]: Can be replaced with just `anti_wps4?? = 'unreadable class'` (sorry, I don't remember the violation number)

File: tests/test_visitors/test_ast/test_conditions/test_useless_len_call.py [hunk 2]
```diff
@@ -5,16 +5,16 @@
 )
 from wemake_python_styleguide.visitors.ast.conditions import IfStatementVisitor
 
-correct_call1 = 'print(len([]))'
+correct_call = 'print(len([]))'
```
Comment (sobolevn) [inline tests/test_visitors/test_ast/test_conditions/test_useless_len_call.py]: I have a question about this case.

What do you think:
1. Should we allow `[a-z]\d+$` pattern? (1 and other numbers can always be located at the end of the variable name)
2. Should we stick to this implementation?

I, personally, like the first one more.

File: tests/test_visitors/test_filenames/test_module/test_module_redable_name.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    """Testing that unreadable characters combinations do not allowed."""
+    visitor = WrongModuleNameVisitor(default_options, filename=filename)
+    visitor.run()
+
+    assert_errors(visitor, [UnreadableNameViolation])
+
+
+@pytest.mark.parametrize('filename', [
+    'ordinary',
+    'first_module',
```
Comment (sobolevn) [inline tests/test_visitors/test_filenames/test_module/test_module_redable_name.py]: We should also test `TestO_0`: is it valid?

File: wemake_python_styleguide/constants.py [hunk 4]
```diff
@@ -379,3 +379,15 @@
 
 # Used to specify as a placeholder for `__init__`:
 INIT: Final = '__init__'
+
+# Used to specify unreadable combinations of letters and letters with numbers
+UNREADABLE_CHARACTER_COMBINATIONS = frozenset((
```
Comment (sobolevn) [inline wemake_python_styleguide/constants.py]: It should be annotated with `: Final` and documented in the public section with `#:`

File: wemake_python_styleguide/logic/naming/logical/contain.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+    True
+
+    >>> does_contain_unreadable_characters('0Operations', ['0O'])
+    True
+
+    >>> does_contain_unreadable_characters('O0', ['O0'])
+    True
+
+    """
+    for unreadable_combination in character_combinations:
```
Comment (sobolevn) [inline wemake_python_styleguide/logic/naming/logical/contain.py]: Let's replace this logic with `any(... for unreadable_combination in character_combinations )`

File: wemake_python_styleguide/logic/tree/imports.py [hunk 6]
```diff
@@ -1,7 +1,9 @@
 from typing import List
 
 from wemake_python_styleguide import constants
-from wemake_python_styleguide.logic.naming import logical
+from wemake_python_styleguide.logic.naming.logical import (  # noqa: I001
```
Comment (sobolevn) [inline wemake_python_styleguide/logic/tree/imports.py]: Sorry, `I001` is not allowed to be ignored. This gave me an idea for a new rule! Thank you! https://github.com/wemake-services/wemake-python-styleguide/issues/1248

File: wemake_python_styleguide/violations/naming.py [hunk 7]
```diff
@@ -779,3 +781,39 @@ class WrongUnusedVariableNameViolation(ASTViolation):
...
     code = 123
+
+
+@final
+class UnreadableNameViolation(MaybeASTViolation):
+    """
+    Forbids to have variable or module names which could be difficult to read.
+
+    Reasoning:
+        Currently one can name your classes like so: `ControlIn`
```
Comment (sobolevn) [inline wemake_python_styleguide/violations/naming.py]: We use double back quotes for formatting.
```
``
```

File: wemake_python_styleguide/violations/naming.py [hunk 8]
```diff
@@ -779,3 +781,39 @@ class WrongUnusedVariableNameViolation(ASTViolation):
...
+        Inside it is just L and i, but we cannot tell it from the word.
```
Comment (sobolevn) [inline wemake_python_styleguide/violations/naming.py]: Can you please also format individual letters here with backquotes?

File: wemake_python_styleguide/violations/naming.py [hunk 9]
```diff
@@ -779,3 +781,39 @@ class WrongUnusedVariableNameViolation(ASTViolation):
...
+        There are a lot other combinations which are unreadable
```
Comment (sobolevn) [inline wemake_python_styleguide/violations/naming.py]: Missing end dot.

File: wemake_python_styleguide/violations/naming.py [hunk 10]
```diff
@@ -779,3 +781,39 @@ class WrongUnusedVariableNameViolation(ASTViolation):
...
+
+    Solution:
+        We need to forbid unreadable letter combinations.
```
Comment (sobolevn) [inline wemake_python_styleguide/violations/naming.py]: The solution is: `Rename your entity not to contain unreadable sequences.`

File: wemake_python_styleguide/violations/naming.py [hunk 11]
```diff
@@ -779,3 +781,39 @@ class WrongUnusedVariableNameViolation(ASTViolation):
...
+
+    This rule checks: modules, variables, attributes,
```
Comment (sobolevn) [inline wemake_python_styleguide/violations/naming.py]: We usually put this before `Reasoning`

Comment (sobolevn) [reply wemake_python_styleguide/violations/naming.py]: You also have to add a link to the constant with all letters. See how we use `:py:data:` for this.

File: wemake_python_styleguide/visitors/ast/naming.py [hunk 12]
```diff
@@ -19,15 +19,19 @@
...
 )
 from wemake_python_styleguide.logic import nodes
 from wemake_python_styleguide.logic.naming import (
     access,
     blacklists,
     builtins,
-    logical,
     name_nodes,
 )
+from wemake_python_styleguide.logic.naming.logical import (  # noqa: I001
```
Comment (sobolevn) [inline wemake_python_styleguide/visitors/ast/naming.py]: The same about `I001`

File: wemake_python_styleguide/visitors/ast/naming.py [hunk 13]
```diff
@@ -74,20 +78,29 @@ def check_name(
...
 
         if not is_first_argument:
-            if logical.is_wrong_name(name, SPECIAL_ARGUMENT_NAMES_WHITELIST):
+            if name_check.is_wrong_name(name, SPECIAL_ARGUMENT_NAMES_WHITELIST):
                 self._error_callback(
                     naming.ReservedArgumentNameViolation(node, text=name),
                 )
 
-        if logical.does_contain_unicode(name):
+        if contain.does_contain_unicode(name):
```
Comment (sobolevn) [inline wemake_python_styleguide/visitors/ast/naming.py]: Let's use `alphabet`  instead of `contain`. How do you like it?

File: wemake_python_styleguide/visitors/filenames/module.py [hunk 14]
```diff
@@ -1,7 +1,11 @@
 from typing_extensions import final
 
 from wemake_python_styleguide import constants
-from wemake_python_styleguide.logic.naming import access, logical
+from wemake_python_styleguide.logic.naming import access
+from wemake_python_styleguide.logic.naming.logical import (  # noqa: I001
```
Comment (sobolevn) [inline wemake_python_styleguide/visitors/filenames/module.py]: The same about `I001`

File: wemake_python_styleguide/visitors/filenames/module.py [hunk 15]
```diff
@@ -31,9 +36,10 @@ def visit_filename(self) -> None:
         self._check_module_name()
         self._check_module_name_length()
         self._check_module_name_pattern()
+        self._check_module_name_readability()
 
     def _check_module_name(self) -> None:
-        if logical.is_wrong_name(self.stem, constants.MODULE_NAMES_BLACKLIST):
+        if name_check.is_wrong_name(self.stem, constants.MODULE_NAMES_BLACKLIST):  # noqa: E501
```
Comment (sobolevn) [inline wemake_python_styleguide/visitors/filenames/module.py]: sorry, `E501` cannot be ignored.

File: tests/test_visitors/test_ast/test_conditions/test_useless_len_call.py (hunk 2 above)
Comment (Kvm99) [reply tests/test_visitors/test_ast/test_conditions/test_useless_len_call.py]: Should not we check only `'l1+$` because other cases correct by default?

Comment (sobolevn) [reply tests/test_visitors/test_ast/test_conditions/test_useless_len_call.py]: Yes, you are right.

Comment (sobolevn) [reply tests/test_visitors/test_ast/test_conditions/test_useless_len_call.py]: Oh, one more: `O0`

Comment (Kvm99) [reply tests/test_visitors/test_ast/test_conditions/test_useless_len_call.py]: I suggest to change WPS cases like correct_call1 to correct_call_1 and do not allow any `(l1|O0)+$`. It seems strange to me allow smth in one case and disallow in other

File: tests/test_visitors/test_filenames/test_module/test_module_redable_name.py (hunk 3 above)
Comment (Kvm99) [reply tests/test_visitors/test_filenames/test_module/test_module_redable_name.py]: `TestO_0` has WrongModuleNamePatternViolation and UnderscoredNumberNameViolation. Why we should check it here if our unreadable list it:
```'lI',
    'Il',
    '1l',
    'l1',
    '1I',
    '1i',
    '0O',
    'O0'

File: tests/test_visitors/test_ast/test_conditions/test_useless_len_call.py (hunk 2 above)
Comment (sobolevn) [reply tests/test_visitors/test_ast/test_conditions/test_useless_len_call.py]: > It seems strange to me allow smth in one case and disallow in other

I agree, we value consistency. 
But, `correct_call_1` is not allowed due to our underscore policies. 

I would still allow `l1$` because it feels like a popular pattern people should be aware about.

File: tests/test_visitors/test_filenames/test_module/test_module_redable_name.py (hunk 3 above)
Comment (sobolevn) [reply tests/test_visitors/test_filenames/test_module/test_module_redable_name.py]: Because it is a corner case that is related to underscore name patterns.

File: tests/test_visitors/test_ast/test_conditions/test_useless_len_call.py (hunk 2 above)
Comment (Kvm99) [reply tests/test_visitors/test_ast/test_conditions/test_useless_len_call.py]: Also see UnderscoredNumberNameViolation. But still think that unreadable is still unreadable even in the end of name

Comment (sobolevn) [reply tests/test_visitors/test_ast/test_conditions/test_useless_len_call.py]: @AlwxSin what do you think? You have edited this rule recently, maybe you have some clever way to solve it.

File: wemake_python_styleguide/violations/naming.py (hunk 11 above)
Comment (Kvm99) [reply wemake_python_styleguide/violations/naming.py]: > We usually put this before `Reasoning`

My code is based on examples I found: https://github.com/wemake-services/wemake-python-styleguide/blob/master/wemake_python_styleguide/violations/naming.py#L328

Comment (sobolevn) [reply wemake_python_styleguide/violations/naming.py]: Ok! I would later need to fix that by adding explicit paragraph ordering. Thanks for the find 👍

File: wemake_python_styleguide/logic/tree/imports.py (hunk 6 above)
Comment (Kvm99) [reply wemake_python_styleguide/logic/tree/imports.py]: I had strange problems with order of imports. However, `isort` could help me

Comment (sobolevn) [reply wemake_python_styleguide/logic/tree/imports.py]: What kind of problems? Do you want to report a new issue?

**Comment by sobolevn** [comment]
Thanks a lot, @Kvm99!
